third_party/move/move-vm/runtime/src/data_cache.rs

```diff
@@ -9,40 +9,104 @@ use crate::{
         module_storage::FunctionValueExtensionAdapter,
         ty_layout_converter::LayoutConverter,
     },
-    ModuleStorage,
+    ModuleStorage, RuntimeEnvironment,
 };
 use bytes::Bytes;
 use move_binary_format::errors::*;
 use move_core_types::{
     account_address::AccountAddress,
     effects::{AccountChanges, ChangeSet, Changes},
     gas_algebra::NumBytes,
+    identifier::IdentStr,
     language_storage::{StructTag, TypeTag},
     value::MoveTypeLayout,
     vm_status::StatusCode,
 };
 use move_vm_types::{
-    gas::DependencyGasMeter,
+    gas::{DependencyGasMeter, DependencyKind, GasMeter},
     loaded_data::runtime_types::Type,
-    resolver::ResourceResolver,
+    resolver::{ResourceResolver, ResourceSizeInfo},
     value_serde::{FunctionValueExtension, ValueSerDeContext},
     values::{GlobalValue, Value},
+    views::TypeView,
 };
 use std::collections::btree_map::{BTreeMap, Entry};
 
+struct TypeWithRuntimeEnvironment<'a, 'b> {
+    ty: &'a Type,
+    runtime_environment: &'b RuntimeEnvironment,
+}
+
+impl TypeView for TypeWithRuntimeEnvironment<'_, '_> {
```

Comment:
This is the same as in interpreter? Maybe it is worth making this thing `pub(crate)` and share across interpreter file and here